airqo-platform/AirQo-frontend
# [WEBSITE] Display AirQlouds based on Users country/location

#### Summary of Changes (What does this PR do?)
- This PR modifies the locationTracker component to use the OpenCage API for geolocation. This API offers more requests per day (2500) than the IPAPI, which has a limit of 1000 requests per day. The OpenCage API also has a free developer tier that is fairly suitable for production.
- The locationTracker component has been refactored to handle API errors and failed responses gracefully. It also avoids unnecessary API calls by using both session storage and local storage accordingly since we are using a free tier with limited requests.
- Additionally, a custom popup modal has been implemented to enhance the user interface and user experience. This modal appears after a delay of 5000 milliseconds, or 5 seconds, providing a more engaging user experience.
- The user location is tracked once and saved in session and localStorage to avoid repeated API calls to track the users location.
- A new .env variable has been added to store the OpenCage API key. The key was obtained using my GitHub account, but I suggest we use the company account to get a new key for AirQo.
- For more details on the sources used in this PR, please visit the following links:
   - OpenCage Data https://opencagedata.com/api
   - Geocode Maps https://geocode.maps.co/
   - IPAPI  https://ipapi.co/#api

#### NOTICE TO VIEW THE CHANGES WORKING
- The new .env variable should be added to secrets accordingly.

#### NB:
- If the user’s browser is set to accept location tracking by default, the user’s location will be tracked in the background without displaying a pop-up notification.
#### Status of maturity (all need to be checked before merging):


#### How should this be manually tested?
- Load the website: Start by loading the website on your browser. If you haven’t yet accepted the location tracking request, a popup should appear asking for your permission.
- Check the network tab: Clear previous requests in your network tab, then click ‘Yes’ on the location tracking request. You should see a response from the OPENCAGE API endpoint.
- Deny permission: If you deny permission to track your location during a session, the request modal won’t pop up again for that session. However, if you open the website in a new session and haven’t accepted location tracking, you will be asked again.
- Check AirQlouds display: Once your location is tracked, the AirQlouds for your country will be displayed in the footer.
- Use a VPN: You can also test this feature by starting your VPN to change your location. Please note that if your current location (as determined by the VPN) is not on the list of countries where AirQlouds are available, it will default to displaying Uganda’s AirQloud.

- [WEB-409](https://airqoteam.atlassian.net/browse/WEB-406)

#### Screenshots (optional)
![image](https://github.com/airqo-platform/AirQo-frontend/assets/97807374/0562bf00-7069-4c06-a02f-c2f54176e45b)
![image](https://github.com/airqo-platform/AirQo-frontend/assets/97807374/dd472b27-4c98-4fb8-b1d4-1c6aec4e5b31)
![Screenshot 2023-10-12 042108](https://github.com/airqo-platform/AirQo-frontend/assets/97807374/c4c9a39d-1239-4f1d-baa1-d6c9dd893fe6)


[WEB-409]: https://airqoteam.atlassian.net/browse/WEB-409?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
